mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java

```diff
@@ -171,117 +191,209 @@ public void commitJob(JobContext originalContext) throws IOException {
+   */
+  private static ExecutorService fileExecutor(Configuration conf) {
+    int size = conf.getInt(InputFormatConfig.COMMIT_FILE_THREAD_POOL_SIZE,
+        InputFormatConfig.COMMIT_FILE_THREAD_POOL_SIZE_DEFAULT);
+    return Executors.newFixedThreadPool(
+        size,
+        new ThreadFactoryBuilder()
+            .setDaemon(true)
+            .setPriority(Thread.NORM_PRIORITY)
+            .setNameFormat("iceberg-commit-file-pool-%d")
+            .build());
+  }
+
+  /**
+   * Executor service for parallel handling of table manipulation. Could return null, if no parallelism is possible.
+   * @param conf The configuration containing the pool size
+   * @param maxThreadNum The number of requests we want to handle (might be decreased further by configuration)
+   * @return The generated executor service, or null if executor is not needed.
+   */
+  private static ExecutorService tableExecutor(Configuration conf, int maxThreadNum) {
+    int size = conf.getInt(InputFormatConfig.COMMIT_TABLE_THREAD_POOL_SIZE,
+        InputFormatConfig.COMMIT_TABLE_THREAD_POOL_SIZE_DEFAULT);
+    size = Math.min(maxThreadNum, size);
+    if (size > 1) {
+      return Executors.newFixedThreadPool(
+          size,
+          new ThreadFactoryBuilder()
+              .setDaemon(true)
+              .setPriority(Thread.NORM_PRIORITY)
+              .setNameFormat("iceberg-commit-table-pool-%d")
+              .build());
+    } else {
+      return null;
+    }
   }
 
   /**
    * Get the data committed data files for this job.
+   * @param location The location of the table
    * @param jobContext The job context
    * @param io The FileIO used for reading a files generated for commit
    * @param throwOnFailure If <code>true</code> then it throws an exception on failure
    * @return The list of the committed data files
    */
-  private static List<DataFile> dataFiles(JobContext jobContext, FileIO io, boolean throwOnFailure) {
+  private static List<DataFile> dataFiles(ExecutorService executor, String location, JobContext jobContext, FileIO io,
+      boolean throwOnFailure) {
     JobConf conf = jobContext.getJobConf();
     // If there are reducers, then every reducer will generate a result file.
     // If this is a map only task, then every mapper will generate a result file.
     int expectedFiles = conf.getNumReduceTasks() > 0 ? conf.getNumReduceTasks() : conf.getNumMapTasks();
 
-    ExecutorService executor = null;
-    try {
-      // Creating executor service for parallel handling of file reads
-      executor = Executors.newFixedThreadPool(
-          conf.getInt(InputFormatConfig.COMMIT_THREAD_POOL_SIZE, InputFormatConfig.COMMIT_THREAD_POOL_SIZE_DEFAULT),
-          new ThreadFactoryBuilder()
-              .setDaemon(true)
-              .setPriority(Thread.NORM_PRIORITY)
-              .setNameFormat("iceberg-commit-pool-%d")
-              .build());
+    List<DataFile> dataFiles = Collections.synchronizedList(new ArrayList<>());
```

Comment:
We usually prefer `ConcurrentLinkedQueue` to `synchronizedList` because it can relax guarantees and gain some performance.

Reply:
Done